### PR TITLE
Updates to adapter training

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

Parallel adapter did not perform as well initially. 

**Collection**: [NLP]

# Changelog 
- Config now lets you select the type of initialization for ColumnLinear.

# Usage
* See  #4717

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml [hunk 1]
```diff
@@ -103,9 +103,12 @@ model:
     answer_field: "target"
 
   adapter_tuning:
+    type: 'parallel'
     adapter_dim: 50
     adapter_dropout: 0.1
+    column_init: 'xavier'
     norm_position: 'pre' # This can be set to 'pre' or 'post', 'pre' is normally what is used.
+    norm_type: 'mixedfused'
```
Comment (MaximumEntropy) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml:111]: This is a generic prefix for both layernorm and rmsnorm can it be more verbose?

Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml:111]: We use the same argument name elsewhere to provide either `layernorm` or `rmsnorm` as possible values. This will probably who use both confused. Can you also include the list of acceptable values for this arg?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml [hunk 2]
```diff
@@ -103,9 +103,12 @@ model:
     answer_field: "target"
 
   adapter_tuning:
+    type: 'parallel'
```
Comment (MaximumEntropy) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml:106]: Can you add a list of possible args here? What does parallel mean?

Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml:106]: If it is meant to differentiate between megatron adapters and vanilla pytorch adapters I would use that nomenclature instead.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py [hunk 3]
```diff
@@ -61,18 +62,30 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
                 )
 
         logging.info(f'Before adding adapters:\n{self.frozen_model.summarize()}')
+
+        if cfg.adapter_tuning.type == "parallel":
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py:66]: For backward compatibility with our earlier adapters you can do cfg.adapter_tuning.get('type', default) here.

File: nemo/collections/nlp/modules/common/megatron/parallel_adapters.py [hunk 4]
```diff
@@ -44,6 +44,8 @@ def __init__(
         dim: int,
         activation: str = 'swish',
         norm_position: str = 'post',
+        norm_type: str = 'mixedfused',
+        column_init: str = 'xavier',
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/parallel_adapters.py:48]: Is this column init because it only initialized the columnlinear layer within the adapter with this initialization scheme?

Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/parallel_adapters.py:48]: We have initialization methods and kwargs for layers in `transformer.py` already, would it be possible to repurpose those?

Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/parallel_adapters.py:48]: Also, can we rename it `init_method` rather than just `init`

File: nemo/collections/nlp/modules/common/megatron/parallel_adapters.py [hunk 5]
```diff
@@ -54,9 +56,17 @@ def __init__(
         self.activation = activation_registry[activation]()
         self.norm_position = norm_position
 
-        self.linear_in = ColumnLinear(in_features, dim, bias=False, init_method=init_method_normal(0.2))
+        if column_init == 'xavier':
+            self.linear_in = ColumnLinear(in_features, dim, bias=False)
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/parallel_adapters.py:60]: Probably good to make all args within this adapter configurable? Such as bias and ColumnLinear init method as well?

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Nice improvements! Lets try and align GPT and T5 adapter nomenclature.

**Comment by Davood-M** [comment]
Great update and improvements. Just a minor comment on the position of adapter 2. Can we have adapter 2 at both positions and switch between them using config?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml (hunk 1 above)
Comment (arendu) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml]: Updated to more specific option names.

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml (hunk 2 above)
Comment (arendu) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_adapter_tuning_config.yaml]: parallel was used signify using Column/Row Linear Parallel instead of nn.Linear. Its not set to 'linear_adapter' or 'parallel_adapter'

File: nemo/collections/nlp/modules/common/megatron/parallel_adapters.py (hunk 4 above)
Comment (arendu) [reply nemo/collections/nlp/modules/common/megatron/parallel_adapters.py]: Renamed the options.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py (hunk 3 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py]: Done!

File: nemo/collections/nlp/modules/common/megatron/parallel_adapters.py (hunk 5 above)
Comment (arendu) [reply nemo/collections/nlp/modules/common/megatron/parallel_adapters.py:61]: I'd leave bias as is, but init methods for row and column linear units are provided now.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py [hunk 6]
```diff
@@ -51,6 +51,11 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
         assert (
```
Comment (Davood-M) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py:51]: Not sure if important, but should we also add a validation step to check if all necessary keys are present in the given config (like adapter_dim or dropout)?

**Comment by arendu** [comment]
> Just a minor comment on the position of adapter 2. Can we have adapter 2 at both positions and switch between them using config?
The current position is the canonical position per the paper im following. I had moved it before because of some instability issues, which appear to have been resolved with ptl1.7

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py (hunk 6 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py:51]: dropout is not required, but I can add one for adapter_dim.

**Review by Davood-M — APPROVED** [review]
